### ISSUE TITLE
BUG: read_file() with Fiona fails to read GeoJSON string what includes hash in a property value

### ISSUE THREAD

**Issue by kfischer-hydrosat** [geopandas/geopandas#3344]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [x] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import json
import geopandas as gpd

feature = {'features': {'geometry': {'coordinates': [[[[5.7807100168363, 52.568150742083, 0.0], [5.7751310220853, 52.567061692737, 0.0], [5.7758176675931, 52.565750884908, 0.0], [5.781375204672, 52.566853009536, 0.0], [5.7807100168363, 52.568150742083, 0.0]]]], 'type': 'MultiPolygon'}, 'properties': {'name': 'TB #8'}, 'type': 'Feature'}}

feature_gdf = gpd.read_file(json.dumps(feature), engine="fiona")

```

#### Problem description
The above feature cannot be read with geopandas 0.14.4. The output I'm receiving is as follows:

```
Traceback (most recent call last):
  File "fiona\\ogrext.pyx", line 136, in fiona.ogrext.gdal_open_vector
  File "fiona\\_err.pyx", line 291, in fiona._err.exc_wrap_pointer
fiona._err.CPLE_OpenFailedError: Failed to read GeoJSON data
```
When removing the ```#``` in the name, the feature is read without any issues. 

#### Expected Output
The feature is correctly read as a GeoDataFrame.

#### Output of ``geopandas.show_versions()``

<details>

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.14.4
numpy      : 1.26.4
pandas     : 2.2.2
pyproj     : 3.6.1
shapely    : 2.0.4
fiona      : 1.9.6
geoalchemy2: None
geopy      : None
matplotlib : 3.8.4
mapclassify: 2.6.1
pygeos     : None
pyogrio    : None
psycopg2   : None
pyarrow    : None
rtree      : 1.2.0

</details>

**Comment by brendan-ward** [comment]
Indeed, this is an error in `geopandas.io.util.py::vsi_path` handling of hashes in the filename.  It looks like in order to avoid parsing the string passed as the filename as a path (that may include hashes like a URL), we need to detect that the input is actually GeoJSON and not a path.

Fiona reads this fine, but we parse the path prior to calling Fiona, which causes this to break.


As a workaround, pyogrio reads this fine:

```Python
>>> feature_gdf = gpd.read_file(json.dumps(feature), engine="pyogrio")
>>> feature_gdf
    name                                           geometry
0  TB #8  MULTIPOLYGON Z (((5.78071 52.56815 0.00000, 5....
```


If you instead either write the GeoJSON to disk and read from there, or make the the input a file-like (with a `.read()` method), it will work in Fiona as well:

```Python
>>> from io import StringIO
>>> gpd.read_file(StringIO(json.dumps(feature)), engine='fiona')
    name                                           geometry
0  TB #8  MULTIPOLYGON Z (((5.78071 52.56815 0.00000, 5....
```

I'm actually surprised that pyogrio / Fiona handle this fine, as it seems uncommon to pass in a GeoJSON string as a filename, but apparently GDAL detects that and correctly reads it.

**Comment by JamesParrott** [comment]
I don't know if it helps, but iyour code doesn't error for me using fiona==1.9.5 and geopandas 0.14.1.

**Comment by kfischer-hydrosat** [comment]
With the following, it also works for me:

```
GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.12.1
GEOS lib   : None
GDAL       : 3.9.0
GDAL data dir: None
PROJ       : 9.4.0
PROJ data dir: C:\ProgramData\miniforge3\envs\irriwatch-engine\Library\share\proj
PYTHON DEPENDENCIES
-------------------
geopandas  : 0.14.3
numpy      : 1.23.5
pandas     : 1.4.4
pyproj     : 3.6.1
shapely    : 2.0.4
fiona      : 1.9.6
geoalchemy2: None
geopy      : None
matplotlib : 3.8.4
mapclassify: 2.6.1
pygeos     : None
pyogrio    : None
psycopg2   : 2.9.9 (dt dec pq3 ext lo64)
pyarrow    : 16.1.0
rtree      : 1.2.0
```

I believe the bug was introduced with 0.14.4 only.

**Comment by jorisvandenbossche** [comment]
> I believe the bug was introduced with 0.14.4 only.

That's correct. This was introduced by https://github.com/geopandas/geopandas/pull/3232, which is only included in 0.14.4 (and in upcoming 1.0).

**Comment by jorisvandenbossche** [comment]
And in the meantime also already fixed through https://github.com/geopandas/geopandas/pull/3280. So this issue will be resolved with the upcoming 1.0 release (planned for early next week). We could also consider this for a 0.14.5 release.

And while I am also surprised this actually works with passing the json string, given this has been working and is being used, it would be good to add a test for this to avoid breaking it again in the future.

**Comment by m-richards** [comment]
Since this is resolved in 1.0 and a regression test was added in #3347 this can be closed.
